Store metadata and BIOM data in a compressed format in the HTML / JavaScript

fedarko opened the pull request:
Closes #158.

The bulk of the changes here are in the new `compression_utils` python module (which gets compressed data ready to send to jinja2), and in the JavaScript BIOM table code (the compressed BIOM / sample metadata is formatted pretty differently from the previous way this data was stored, so I ended up rewriting most of the BIOM code -- the existing JS BIOM tests made this much less painful than I thought it might be :). I also added a decent amount of tests for the above modules, and made a few other changes throughout the codebase.

In terms of size compression, the early result from [here](https://github.com/biocore/empress/issues/158#issuecomment-649905429) still holds on the HMP dataset: the empress.html file goes down from 47MB to 26MB. The loading times seem... maybe slightly faster? but still mostly the same, at least for this dataset. Should probably benchmark this with an even larger dataset, though.

ElDeveloper commented on empress/compression_utils.py:
```diff
@@ -0,0 +1,299 @@
+def compress_table(table):
```
Can you add a copyright header?

fedarko replied:
Done, thanks!

ElDeveloper commented on empress/compression_utils.py:
```diff
@@ -0,0 +1,299 @@
+        f_ids_to_indices: dict
+            Inverse of f_ids: this maps feature IDs to their indices in f_ids,
+            analogous to s_ids_to_indices.
+        compressed_table: list
+            Two-dimensional list. The "outer list" is of length len(s_ids).
+            Each position i within this outer list holds an "inner list" of
+            arbitrary (but within the range [1, len(f_ids)]) length.
+            The i-th inner list contains the feature indices of the
+            features present (i.e. at any abundance > 0) within the
+            sample with index i. Each inner list is sorted in ascending order.
```
Is there any easy way to keep the abundance data around? his will likely be helpful for other issues down the road.

fedarko replied:
Not sure about "easy" ways. The simplest solution that comes to mind is storing abundance data in a separate structure with the same dimensions as `compressed_table` -- so if a sample's row in the table is

`[0, 5, 6, 100]`

indicating that that sample contains four features at those indices, then this sample's row in the abundance data structure will be the corresponding abundances for those features, e.g.

`[1, 2, 20, 3]`

or something like that. However, I think this will require a pretty significant amount of work to squeeze in, so my vote is in favor of delaying that to a future issue. (I'll cc @antgonza here in case he has any thoughts on this.)

ElDeveloper replied:
That sounds like an easy enough solution, thanks! Let's not implement that but whenever needed this could be expanded as you describe.

ElDeveloper commented on empress/support_files/js/biom-table.js:
```diff
@@ -1,294 +1,447 @@
+        // Now, we can go through the table and find samples with matches
+        var containingSampleIDs = [];
+        _.each(this._tbl, function (presentFeatureIndices, sIdx) {
+            var sampleHasMatch = _.some(fIndices, function (fIdx) {
```
Can you move this anonymous function o outside of the _.each call?

fedarko replied:
Done, thanks!

fedarko commented on empress/compression_utils.py:
```diff
@@ -0,0 +1,299 @@
+    """
+    sample_ids = s_ids_to_indices.keys()
+    num_shared_samples = len(metadata.index.intersection(sample_ids))
+    if num_shared_samples < len(sample_ids):
+        # Sanity check: metadata's samples should be a superset of the samples
+        # in s_ids_to_indices (empty samples were removed from the
+        # latter during table compression). If the metadata's samples are
+        # instead *missing* samples that are in s_ids_to_indices,
+        # something is seriously wrong.
+        raise ValueError(
+            "Metadata is missing sample IDs in s_ids_to_indices."
+        )
```
Kind of -- at this point matching, etc. should already have been done, so we know that the data should be sane. This check is just verifying that the `s_ids_to_indices` mapping was constructed in `compress_table()` (or wherever we got it from) properly.

That said, this error should never be thrown in practice, but if the core.py code changes and breaks something in how this function is called then this error will indicate that something is up. (Esp. if that change happens, say, a few months - a year from now when we've all forgotten how this code works...) I can remove this check if you really want me to, but I would strongly prefer to keep it if possible -- it shouldn't be a major performance bottleneck, and the peace of mind from having each step of this use sanity checking is desirable IMO.

ElDeveloper replied:
If there isn't a test case that would flag a situation like that, then we should definitely add that. But otherwise I think we would still be able to catch those errors, right?

My concern is not performance, more along the lines of redundancy and when is enough redundancy sufficient.

fedarko commented on empress/compression_utils.py:
```diff
@@ -0,0 +1,299 @@
+            "Metadata is missing sample IDs in s_ids_to_indices."
+        )
+
+    if sorted(s_ids_to_indices.values()) != list(range(len(sample_ids))):
+        raise ValueError("Indices (values) of s_ids_to_indices are invalid.")
```
Same point as with https://github.com/biocore/empress/pull/233#discussion_r451038304 -- this should never happen in practice because `compress_table()` is well-tested, but it's just an added layer of security. As with there, I don't *think* sorting a list of numbers will be a huge performance issue, so I would prefer to keep this if possible.

ElDeveloper replied:
In principle we aren't exposing compression_utils to 3rd parties so the functionality should be mostly within Empress. If this was to be exposed to 3rd parties then I can understand why you would want to add these checks. Are there hotspots in the code that you think aren't being tested enough that would lead to issues like this?

fedarko commented on empress/compression_utils.py:
```diff
@@ -0,0 +1,299 @@
+    for i in range(len(sample_ids)):
+        metadata_vals.append(indices_to_metadata_vals[i])
+
+    return [str(c) for c in str_metadata.columns], metadata_vals
```
This is a good point -- I just refactored the function to do this. It relies a lot more on pandas builtins so it's simpler now :)

fedarko commented on empress/compression_utils.py:
```diff
@@ -0,0 +1,299 @@
+        # We don't return anything -- this function's only purpose is to cause
+        # side effects within the scope of compress_table(). This is... kind
+        # of a cursed way of using df.apply() but it works, and it should be
+        # decently efficient since .apply() is faster than most iteration \._./
```
I would prefer to keep this using `DataFrame.apply()`, since that is usually much faster than normal Python for loops when iterating over large DataFrames ([ref](https://engineering.upside.com/a-beginners-guide-to-optimizing-pandas-code-for-speed-c09ef2c6a4d6)). (We had a similar discussion a few weeks ago [about this code](https://github.com/biocore/empress/pull/183/files#r442490052), and apply proved to be much faster than manual iteration there.)

However, I agree that the function used for applying had too many side effects. I refactored this part of `compress_table()` so that the inner apply function only modifies `compressed_table` and nothing else -- this should make the code much easier to follow while still letting us get the speed benefits of using apply. LMK if this seems like a fair compromise to you.

fedarko commented:
@ElDeveloper I've addressed all of the comments (with the exception of the two discussions on the checks at the start of `compress_sample_metadata()`) and merged in the changes from upstream, so conflicts should be resolved. Please let me know if you'd like to discuss things further tomorrow so we can get this in for the release!

ElDeveloper commented:
Thanks @fedarko!
